src/ucx_plugin.c

```diff
@@ -5,42 +5,580 @@
  * See LICENSE.txt for license information
  ************************************************************************/
 
+#include <pthread.h>
 #include <stdint.h>
+#include <sys/time.h>
+#include <unistd.h>
 
+#include "core.h"
+#include "ibvwrap.h"
 #include "nccl.h"
 #include "nccl_net.h"
-#include "debug.h"
+#include "param.h"
+#include "socket.h"
+#include "ucp/api/ucp.h"
+
+#define UCXCHECK(cmd) do {                           \
+  int e = cmd;                                       \
+  if( UCS_OK != e ) {                                \
+    WARN("Failed: UCX error %s:%d '%d' %s\n",        \
+        __FILE__,__LINE__, e, ucs_status_string(e)); \
+    return ncclInternalError;                        \
+  }                                                  \
+} while(0)
+
+#define UCXCHECK_VOID(cmd) do {                      \
+  int e = cmd;                                       \
+  if( UCS_OK != e ) {                                \
+    WARN("Failed: UCX error %s:%d '%d' %s\n",        \
+        __FILE__,__LINE__, e, ucs_status_string(e)); \
+  }                                                  \
+} while(0)
+
+NCCL_PARAM(UCXDisable, "UCX_DISABLE", 0);
 
 extern ncclDebugLogger_t pluginLogFunction;
+static const ucp_tag_t tag      = 0xABADBABE;
+static const ucp_tag_t tag_mask = 0xFFFFFFFFFFFFFFFF;
 
-ncclResult_t nccl_ucx_init(ncclDebugLogger_t logFunction) {
-  WARN("NET/UCX: not implemented");
-  return ncclInternalError;
-}
+static int ncclNIbDevs = -1;
+extern int ncclNSharpDevs;
+
+typedef struct ucx_mhandle {
+  ucp_mem_h  ucp_memh;
+  ucp_rkey_h rkey;
+} ucx_mhandle_t;
+
+struct ncclIbDev ncclIbDevs[MAX_IB_DEVS];
+struct userIbDev userIbDevs[MAX_IB_DEVS];
 
 ncclResult_t nccl_ucx_devices(int* ndev) {
-  WARN("NET/UCX: not implemented");
-  return ncclInternalError;
+  *ndev = ncclNIbDevs;
+  return ncclSuccess;
 }
 
-ncclResult_t nccl_ucx_get_properties(int dev, ncclNetProperties_t* props) {
-  WARN("NET/UCX: not implemented");
-  return ncclInternalError;
+ncclResult_t nccl_ucx_pci_path(int dev, char** path) {
```

Comment:
should we move common functions for ib and ucx into common file (p2p_plugin.c)?

Reply:
did you use this func? this file also has ncclIbGetPciPath

Reply:
Removed this function